xmodule/docs/decisions/0004-evaluate-wasm-safe-exec.rst

```diff
@@ -0,0 +1,107 @@
+
+Cross-Platform Support
+----------------------
+
+AppArmor only works in Linux, and cannot run on Linux guest Docker containers on a macOS host machine. This makes it difficult for developers running macOS to develop and test features related to sandboxed code execution. WebAssembly can run on any major operating system.
+
+Decisions
+*********
+
+#. We will create a new library in a new repo (``wasm-safe-exec``) for safely executing instructor-authored code using WebAssembly as the execution mechanism.
+
+   * The library will use ``wasmtime-py`` for Wasm code execution.
+   * The library must be runnable in edx-platform, meaning that it still needs to support Python 3.8 until edx-platform has been upgraded to 3.11.
+   * The library must be able to simultaneously support multiple code execution runtime environments.
+   * The library must not make ``edx-platform`` specific assumptions–e.g. ``UsageKeys``. It should be generally reusable for untrusted code execution by other projects.
+
+#. We will use the WebAssembly Component Model and create a Wasm Interface Type (WIT) file defining the interface between the host library and guest code execution runtime environments.
+
+   * It should be possible for third parties to easily create their own runtime environments.
+
+#. We will create two simple runtimes to start: Python 3.11 and JavaScript ES14.
+
+   * Runtime binaries will be created with `componentize-py <https://component-model.bytecodealliance.org/language-support/python.html>`_ for Python and and `jco <https://component-model.bytecodealliance.org/language-support/javascript.html>`_ for JavaScript.
+   * Python 3.11 is the first version of Python that officially supports compiling to ``wasm32-wasi``, which is why it's the one we'll start with.
+   * The Python runtime must be able to accept a course's ``python_lib.zip`` file.
+   * Parity with codejail is not feasible at this time because only a handful of our popular libraries are available. However, we should include numpy (`using Fermyon's instructions <https://www.fermyon.com/blog/introducing-componentize-py>`_) and whatever pure Python libraries are practical to do so.
+   * Preamble code (e.g. certain common library imports) will be the responsibility of the runtime environments.
+   * For this first iteration, the Wasm runtime binary build process can simply be documented, and does not need to be part of any package or CI. In particular, we do *not* want to add 100 MB+ ``.wasm`` files as a direct dependency of edx-platform.
+
+#. We will add this functionality as an opt-in, experimental feature to edx-platform.
+
+   * There will be a ``CourseWaffleFlag`` toggling the feature as a whole.
+   * The ``<script>`` tags used by the ProblemBlock will be extended with two new optional attributes:
+
+     * ``runtime`` will specify the name of the runtime that the problem execution will be dispatched to.
```

Comment:
It might be interesting to allow course authors to specify something akin to semantic versions -- for example `python3`, `python3.11`, `python3.11-build20231121` (a specific build with specific library versions). `python3` would be cutting edge, whereas `python3.11` would be more stable.

Reply:
So this is something where we can get more user feedback as we go through the process, and I intentionally didn't go into the details of exactly what conventions we'd use because I figured that would be its own ADR. :-) But my current thoughts are:

1. Stability is much more important than being current for content, because we want to lower the content maintenance burden as much as possible.
2. To have content that is portable across instances, we want to have a relatively small number of target runtimes.

Which makes me think that the releases of these runtimes should be infrequent, with no officially supported "cutting edge" versions that are moving targets, like `python3`. I don't know how many runtime stacks we'd want to officially support, but I wouldn't want to make a release more frequently than every couple of years. And I'd only want to patch that release for serious bugs that affect correctness.